### PR TITLE
[enhancement] check that all sklearnex estimators are centrally tested

### PR DESCRIPTION
**Pull request by icfaust**
### Description

This adds ```BasicStatistics``` and ```IncrementalBasicStatistics``` to the ```SPECIAL_INSTANCES``` since they cannot be added easily to the patch_map.  This adds a test to ```tests/test_common.py``` which will check that all estimators which inherit from sklearn's ```BaseEstimator``` without a leading underscore are in either ```PATCHED_MODELS``` or ```SPECIAL_INSTANCES``` such that they are centrally tested via ```sklearnex/tests```

This works by monkeypatching sklearn's ```all_estimators``` which is a function sklearn uses internally to discover all estimators in sklearn.  This is patched to yield all sklearn-style estimators in sklearnex without using a patch map.  This required modifying all sklearn-imported estimators to follow python private variable conventions (leading underscore), which is the bulk of the changes. This is a reasonable change, since we actually would like the sklearn estimators to be private in the various sklearnex modules.

A changes was necessary in ```BasicStatistics``` in order for it to be added to ```SPECIAL_INSTANCES```, where it currently cannot be ```cloned()```, it required naming the ```options``` attribute to ```result_options``` to match the kwarg on ```__init__```. (fixed in a separate PR #2038)

This fixes an issue with ```IncrementalBasicStatistics``` where validate_data is called too often for a fit call.  It follows the conventions of ```IncrementalPCA``` and ```IncrementalEmpericalCovariance``` by adding a check_input kwarg boolean.

This PR has no-performance impact as it adds to testing or is renaming variables.

```BasicStatistics``` is shown by stability testing to not be deterministic, this will be added to the documentation.

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.  
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

**Comment by samir-nasibli** [comment]
@icfaust could you please also include some statistics about the PR change

File: sklearnex/linear_model/logistic_path.py
Comment (samir-nasibli) [inline sklearnex/linear_model/logistic_path.py:1]: We forget to remove this early?

File: sklearnex/tests/test_common.py [hunk 1]
```diff
@@ -52,3 +57,32 @@ def test_target_offload_ban():
...
+            # force root to sklearnex
+            kwargs["path"] = [str(pathlib.Path(__file__).parent.parent)]
+        for pkginfo in func(*args, **kwargs):
+            # Do not allow spmd to be yielded
+            if "spmd" not in pkginfo.name.split("."):
+                yield pkginfo
+    return wrap
+
+
+def test_all_estimators_covered(monkeypatch):
```
Comment (samir-nasibli) [inline sklearnex/tests/test_common.py:79]: Could you please add some docstring to describe the test suit

Comment (icfaust) [reply sklearnex/tests/test_common.py:80]: Done.

File: sklearnex/linear_model/logistic_path.py
Comment (icfaust) [reply sklearnex/linear_model/logistic_path.py:1]: Unnecessarily introduced in #624 , doesn't impact patch map.

**Comment by icfaust** [comment]
> @icfaust could you please also include some statistics about the PR change

All in comparison to last main CI run

| runner | num of tests | sklearnex runtime |
| --- | --- | --- |
| github py3.9sk1.1 lnx | 9086 (+101) | 10 min 15 s (+9 s) |
| github py3.9sk1.1 win | 11763 (+123) | 20 min 13 s (-10 s) |
| github py3.10sk1.2 lnx | 6309 (+71)  | 7 min 31 s (+8 s) | 
| github py3.10sk1.2 win | 6309 (+71) | 13 min 4 s (-54 s)  | 
| github py3.11sk1.3 lnx | 9154 (+101) | 9 min 55 s (+26 s) | 
| github py3.11sk1.3 win | 11853 (+123) | 19 min 41 s (+16 s) |

File: sklearnex/tests/test_common.py [hunk 2]
```diff
@@ -52,3 +57,38 @@ def test_target_offload_ban():
...
+    patch map or covered in special testing via SPECIAL_INSTANCES. The estimator
+    must inherit sklearn's BaseEstimator and must not have a leading underscore.
+    The sklearnex.spmd and sklearnex.preview packages are not tested.
+    """
+    monkeypatch.setattr(pkgutil, "walk_packages", _sklearnex_walk(pkgutil.walk_packages))
+    estimators = all_estimators()
+    print(estimators)
+    for name, obj in estimators:
+        # do nothing if defined in preview
+        if "preview" not in obj.__module__:
```
[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/tests/test_common.py:91]: Preview has not been centrally tested up to this point. It would also conflict with the PATCHED_MODELS, as we would need to then bookkeep for two versions of the same estimator throughout testing.  Individual preview tests are discovered by pytest, but not discovered in a meaningful way by sklearnex.

File: sklearnex/linear_model/logistic_path.py
Comment (samir-nasibli) [inline sklearnex/linear_model/logistic_path.py:1]: I would suggest highlight this in squashed PR commit message this changes or just move this into separate PR

File: sklearnex/tests/test_common.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply sklearnex/tests/test_common.py:91]: We can enable preview via env variable in this case they should be centrally tested for patched models don't they?

**Review by samir-nasibli — APPROVED** [review]
Overall looks good to me! Thank you @icfaust

File: sklearnex/tests/test_common.py [hunk 3]
```diff
@@ -52,3 +57,38 @@ def test_target_offload_ban():
...
```
Comment (Alexsandruss) [inline sklearnex/tests/test_common.py:91]: Why not use estimator as test parameter and pytest.skip preview ones?

File: sklearnex/tests/test_common.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/tests/test_common.py:91]: True, luckily so far we don't do that in any of the CI systems.  I guess it raises questions about what defines a preview estimator, I assumed it was because of code-quality and/or performance.

File: sklearnex/tests/test_common.py (hunk 3 above)
Comment (icfaust) [reply sklearnex/tests/test_common.py:91]: Good question, I would have to move the all_estimators monkeypatch into a fixture and then do an indirect parametrization. Objects like BaseSVM could also show up in the list. It will make sure that multiple failures will both show up, not just the first.  If you want me to do it let me know. @Alexsandruss

Comment (Alexsandruss) [reply sklearnex/tests/test_common.py:91]: @icfaust, yes, it makes same to do it.

Comment (icfaust) [reply sklearnex/tests/test_common.py:91]: Unfortunately using fixtures at collection time is something specifically not supported: https://github.com/pytest-dev/pytest/issues/7140#issuecomment-622237245 In order to do this, I would need to manually monkeypatch instead of using the pytest monkeypatch fixture, which has implications on test isolation. What I will do is change the logic to collect all missing estimators and display them all as a single fail assert.
